spec/requests/uploads_spec.rb

```diff
@@ -0,0 +1,28 @@
+require "rails_helper"
+
+RSpec.describe UploadsController, type: :request do
+  describe "GET #index" do
+    it "renders the new template" do
+      get uploads_path
+      expect(response).to render_template("index")
+    end
+  end
+
+  describe "POST #create" do
+    context "when an invalid file is uploaded" do
+      let(:file) { fixture_file_upload("invalid_file.txt", "text/plain") }
+
+      it "returns an error message" do
+        post uploads_path, params: { oas_file: file }
+        expect(response).to render_template("index")
+      end
+    end
+
+    context "when no file is uploaded" do
+      it "returns an error message" do
+        post uploads_path
+        expect(response).to render_template("index")
+      end
+    end
```

Comment:
[requirement]

The outcome of all three behaviours (root + 2 x contexts) is the same. Therefore, the difference between the three outcomes is not tested. So I'd suggest something like:
```ruby
context "when no file is uploaded" do
  it "returns an error message" do
    post uploads_path
    expect(response).to be_successful
    expect(response.body).to include("No file was selected")
  end
end
```